File: MyProject/Source/MyProject/Private/Game/Libs/Table/TableSys.cpp
#include "MyProject.h"
#include "TableSys.h"
#include "TableBase.h"
#include "TableItemBase.h"
#include "UtilStr.h"
#include "Endian.h"
#include "TableObject.h"
#include "TableCard.h"
#include "TableSkill.h"
#include "TableItemHeader.h"
#include "TableItemBodyBase.h"
#include "ByteBuffer.h"
#include "Common.h"

TableSys::TableSys()
{
    m_dicTable[TableID::TABLE_OBJECT] = new TableBase("ObjectBase_client.bytes", "ObjectBase_client");
    m_dicTable[TableID::TABLE_CARD] = new TableBase("CardBase_client.bytes", "CardBase_client");
    m_dicTable[TableID::TABLE_SKILL] = new TableBase("SkillBase_client.bytes", "SkillBase_client");    // 添加一个表的步骤三

	m_byteBuffer = new ByteBuffer();
}

// 返回一个表
std::vector<TableItemBase*>* TableSys::getTable(TableID::TableID tableID)
{
	TableBase* table = m_dicTable[tableID];
	if (nullptr == table)
	{
		loadOneTable(tableID);
		table = m_dicTable[tableID];
	}
	return table->m_List;
}
		
// 返回一个表中一项，返回的时候表中数据全部加载到 Item 中
TableItemBase* TableSys::getItem(TableID::TableID tableID, uint32 itemID)
{
    TableBase* table = m_dicTable[tableID];
	if (nullptr == table->m_byteBuffer)
	{
		loadOneTable(tableID);
		table = m_dicTable[tableID];
	}
    TableItemBase* ret = TableSys::findDataItem(table, itemID);

	if (nullptr != ret && nullptr == ret->m_itemBody)
    {
        loadOneTableOneItemAll(tableID, table, ret);
    }

    if (nullptr == ret)
    {
		g_pLogSys->log(UtilStr::Format("table name: {0}, table Item {1} 加载失败", (int)tableID, itemID));
    }

	return ret;
}
		
// 加载一个表
void TableSys::loadOneTable(TableID::TableID tableID)
{
	m_byteBuffer->clear();
	m_arrayBuffer.Empty();
	TableBase* table = m_dicTable[tableID];

	FString Filename = FString::Printf(TEXT("%s%s%s%s"), *FPaths::GameContentDir(), TEXT("/Table/"), ANSI_TO_TCHAR(table->m_tableName.c_str()), TEXT(".txt"));

	if (FFileHelper::LoadFileToArray(m_arrayBuffer, *Filename))
	{
		m_byteBuffer->setLength(m_arrayBuffer.GetAllocatedSize());
		m_byteBuffer->writeBytes((char*)(m_arrayBuffer.GetData()), 0, m_arrayBuffer.GetAllocatedSize());
		m_byteBuffer->setPos(0);
		readTable(tableID, m_byteBuffer);
	}
}

// 根据路径查找表的 ID
TableID::TableID TableSys::getTableIDByPath(std::string& path)
{
	TableMapIte beginIte = m_dicTable.begin();
	TableMapIte endIte = m_dicTable.end();
	for(; beginIte != endIte; ++beginIte)
    {
		if (beginIte->second->m_resName == path)
        {
			return beginIte->first;
        }
    }

	return (TableID::TableID)0;
}

// 加载一个表中一项的所有内容
void TableSys::loadOneTableOneItemAll(TableID::TableID tableID, TableBase* table, TableItemBase* itemBase)
{
    if (TableID::TABLE_OBJECT == tableID)
    {
		itemBase->parseBodyByteBuffer<TableObjectItemBody>(table->m_byteBuffer, itemBase->m_itemHeader->m_offset);
    }
    else if (TableID::TABLE_CARD == tableID)
    {
		itemBase->parseBodyByteBuffer<TableCardItemBody>(table->m_byteBuffer, itemBase->m_itemHeader->m_offset);
    }
	else if (TableID::TABLE_SKILL == tableID)  // 添加一个表的步骤四
    {
		itemBase->parseBodyByteBuffer<TableSkillItemBody>(table->m_byteBuffer, itemBase->m_itemHeader->m_offset);
    }
}
		
// 获取一个表的名字
std::string& TableSys::getTableName(TableID::TableID tableID)
{
	TableBase* table = m_dicTable[tableID];
	if (nullptr != table)
	{
		return table->m_tableName;
	}			
	return UtilStr::m_sDefaultStr;
}

// 读取一个表，仅仅读取表头
void TableSys::readTable(TableID::TableID tableID, ByteBuffer* bytes)
{
    TableBase* table = m_dicTable[tableID];
	table->m_byteBuffer = bytes;

	bytes->setEndian(eLITTLE_ENDIAN);
	uint32 len = 0;
	bytes->readUnsignedInt32(len);
    uint32 i = 0;
    TableItemBase* item = nullptr;
    for (i = 0; i < len; i++)
    {
        //if (TableID.TABLE_OBJECT == tableID)
        //{
        //    item = new TableItemObject();
        //}
        item = new TableItemBase();
        item->parseHeaderByteBuffer(bytes);
        // 加载完整数据
        //loadOneTableOneItemAll(tableID, table, item);
        //if (TableID.TABLE_OBJECT == tableID)
        //{
            //item.parseAllByteArray<TableObjectItemBody>(bytes);
        //}
        table->m_List->push_back(item);
    }
}

// 查找表中的一项
TableItemBase* TableSys::findDataItem(TableBase* table, uint32 id)
{
	int size = table->m_List->size();
	int low = 0;
	int high = size - 1;
	int middle = 0;
	uint32 idCur = 0;
			
	while (low <= high)
	{
		middle = (low + high) / 2;
        idCur = (*(table->m_List))[middle]->m_itemHeader->m_uID;
		if (idCur == id)
		{
			break;
		}
		if (id < idCur)
		{
			high = middle - 1;
		}
		else
		{
			low = middle + 1;
		}
	}
			
	if (low <= high)
	{
        return (*(table->m_List))[middle];
	}
	return nullptr;
}

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Table/TableSys.lua
--[[
 * @brief 添加一个表的步骤总共分 4 步
 * // 添加一个表的步骤一
 * // 添加一个表的步骤二
 * // 添加一个表的步骤三
 * // 添加一个表的步骤四
]]

MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "TableSys";
GlobalNS[M.clsName] = M;

function M:ctor()
	self.mDicTable = GlobalNS.new(GlobalNS.MDictionary);
    self.mDicTable:Add(GlobalNS.TableID.TABLE_OBJECT, GlobalNS.new(GlobalNS.TableBase, "ObjectBase_client.bytes", "ObjectBase_client"));
    self.mDicTable:Add(GlobalNS.TableID.TABLE_CARD, GlobalNS.new(GlobalNS.TableBase, "CardBase_client.bytes", "CardBase_client"));
    self.mDicTable:Add(GlobalNS.TableID.TABLE_SKILL, GlobalNS.new(GlobalNS.TableBase, "SkillBase_client.bytes", "SkillBase_client"));    -- 添加一个表的步骤三
    self.mDicTable:Add(GlobalNS.TableID.TABLE_JOB, GlobalNS.new(GlobalNS.TableBase, "proBase_client.bytes", "proBase_client"));
    self.mDicTable:Add(GlobalNS.TableID.TABLE_SPRITEANI, GlobalNS.new(GlobalNS.TableBase, "FrameAni_client.bytes", "FrameAni_client"));
    self.mDicTable:Add(GlobalNS.TableID.TABLE_RACE, GlobalNS.new(GlobalNS.TableBase, "RaceBase_client.bytes", "RaceBase_client"));
    self.mDicTable:Add(GlobalNS.TableID.TABLE_STATE, GlobalNS.new(GlobalNS.TableBase, "StateBase_client.bytes", "StateBase_client"));
end

-- 返回一个表
function getTable(tableID)
	local table = self.mDicTable:value(tableID);
	if (nil == table) then
		self:loadOneTable(tableID);
		table = self.mDicTable:value(tableID);
	end
	return table.mList;
end

-- 返回一个表中一项，返回的时候表中数据全部加载到 Item 中
function M:getItem(tableID, itemID)
    local table = self.mDicTable:value(tableID);
    if (nil == table.mByteBuffer) then
		self:loadOneTable(tableID);
		table = self.mDicTable:value(tableID);
	end
    local ret = self:findDataItem(table, itemID);

    if (nil ~= ret and nil == ret.mItemBody) then
        self:loadOneTableOneItemAll(tableID, table, ret);
    end

    if (nil == ret) then
        -- 日志
    end

	return ret;
end

-- 加载一个表
function M:loadOneTable(tableID)
	local table = self.mDicTable:value(tableID);
	
	local auxBytesLoader = GlobalNS.new(GlobalNS.AuxBytesLoader);
	local path = GlobalNS.UtilPath.CombineTwo(GCtx.mConfig.mPathLst[GlobalNS.ResPathType.ePathTablePath], table.mResName);
	
	auxBytesLoader:syncLoad(path, self, self.onLoadEventHandle);
end

-- 加载一个表完成
function M:onLoadEventHandle(dispObj)
    self.mRes = dispObj;
    if (self.mRes:hasSuccessLoaded()) then
        GCtx.mLogSys:log(self.mRes:getLogicPath(), GlobalNS.LogTypeId.eLogCommon);

        local bytes = self.mRes:getBytes();
        if (nil ~= bytes) then
            self.m_byteArray = GlobalNS.CSSystem.buildByteBuffer();
            self.m_byteArray:clear();
            self.m_byteArray:writeBytes(bytes, 0, bytes.Length, true);
            self.m_byteArray:setPos(0);
            self:readTable(self:getTableIDByPath(self.mRes:getLogicPath()), self.m_byteArray);
        end
    elseif (self.mRes:hasFailed()) then
		GCtx.mLogSys:log(self.mRes:getLogicPath(), GlobalNS.LogTypeId.eLogCommon);
    end

    -- 卸载资源
    self.mRes:dispose();
	self.mRes = nil;
end

-- 根据路径查找表的 ID
function M:getTableIDByPath(path)
	local tablePath = "";
	
    for key, value in pairs(self.mDicTable:getData()) do
		tablePath = GlobalNS.UtilPath.CombineTwo(GCtx.mConfig.mPathLst[GlobalNS.ResPathType.ePathTablePath], value.mResName);
        if (tablePath == path) then
            return key;
        end
    end

    return 0;
end

-- 加载一个表中一项的所有内容
function M:loadOneTableOneItemAll(tableID, table, itemBase)
    if (GlobalNS.TableID.TABLE_OBJECT == tableID) then
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableObjectItemBody
			);
    elseif (GlobalNS.TableID.TABLE_CARD == tableID) then
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableCardItemBody
			);
    elseif (GlobalNS.TableID.TABLE_SKILL == tableID) then  -- 添加一个表的步骤四
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableSkillItemBody
			);
    elseif (GlobalNS.TableID.TABLE_JOB == tableID) then
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableJobItemBody
			);
    elseif (GlobalNS.TableID.TABLE_SPRITEANI == tableID) then
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableSpriteAniItemBody
			);
    elseif (GlobalNS.TableID.TABLE_RACE == tableID) then
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableRaceItemBody
			);
    elseif (GlobalNS.TableID.TABLE_STATE == tableID) then
        itemBase:parseBodyByteBuffer(
			table.mByteBuffer, 
			itemBase.mItemHeader.mOffset, 
			GlobalNS.TableStateItemBody
			);
    end
end

-- 获取一个表的名字
function M:getTableName(tableID)
	local table = self.mDicTable:value(tableID);
	if (nil ~= table) then
		return table.mTableName;
	end
	return "";
end

-- 读取一个表，仅仅读取表头
function M:readTable(tableID, bytes)
    local table = self.mDicTable:value(tableID);
    table.mByteBuffer = bytes;

    --bytes:setEndian(GlobalNS.EEndian.eLITTLE_ENDIAN);
	bytes:setEndian(GlobalNS.CSSystem.EEndian.eLITTLE_ENDIAN);
	--local tmp = nil;
    local count = 0;
    --tmp, count = bytes:readUnsignedInt32(count);
	_, count = bytes:readUnsignedInt32(count);
    local i = 0;
    local item = nil;
    while(i < count) do
        item = GlobalNS.new(GlobalNS.TableItemBase);
        item:parseHeaderByteBuffer(bytes);
        table.mList:Add(item);
		
		i = i + 1
    end
end

-- 查找表中的一项
function M:findDataItem(table, id)
	local size = table.mList:Count();
	local low = 0;
	local high = size - 1;
	local middle = 0;
	local idCur = 0;
	
	while (low <= high) do
		middle = GlobalNS.UtilMath.floor((low + high) / 2);
        idCur = table.mList:at(middle).mItemHeader.mId;
		if (idCur == id) then
			break;
		end
		if (id < idCur)then
			high = middle - 1;
		else
			low = middle + 1;
		end
	end
	
	if (low <= high) then
        return table.mList:at(middle);
	end
	
	return nil;
end

return M;